index.js

```diff
@@ -84,16 +84,36 @@ function typeis (value, types_) {
  * or `content-length` headers set.
  * http://www.w3.org/Protocols/rfc2616/rfc2616-sec4.html#sec4.3
  *
+ * A http/2 request with DataFrame can have no `content-length` header.
+ * https://httpwg.org/specs/rfc7540.html
+ *
+ * A http/2 request without DataFrame send HeaderFrame with end-stream-flag.
+ * If nodejs gets end-stream-flag, then nodejs ends readable stream.
+ * https://github.com/nodejs/node/blob/master/lib/internal/http2/core.js#L301
+ *
  * @param {Object} request
  * @return {Boolean}
  * @public
  */
 
 function hasbody (req) {
-  return req.headers['transfer-encoding'] !== undefined ||
+  return (ishttp2(req) && (req.stream.readable || req.stream._readableState.endEmitted)) ||
```

Comment:
It seems like this still has some kind of issue. I would expect the following test to pass:
```js
      it('should not indicate body', function (done) {
        createBodylessRequest('', function (req) {
          var data = ''
          req.on('data', function (chunk) {
            data += chunk
          })
          req.on('end', function (chunk) {
            process.nextTick(function () {
              assert.strictEqual(typeis.hasBody(req), false)
              done()
            })
          })
        })
      })
```

It seems like just attaching a `data` event will make this module claim the request has a body, even if the original request did not have a body.

Reply:
I addressed issue. I found a parameter `_readableState.sync` which is default true and became false after reading data.

Reply:
Nice! What does that flag actually mean / is it documented so we can rely on some kind of behavior? The underscore prefix just has me worries about the long term.

Reply:
It means 'readable'/'data' event is emitted immediately or not.

See also comment in source.
https://github.com/nodejs/node/blob/master/lib/_stream_readable.js#L104
  // a flag to be able to tell if the event 'readable'/'data' is emitted
  // immediately, or on a later tick.  We set this to true at first, because
  // any actions that shouldn't happen until "later" should generally also
  // not happen before the first read call.

Reply:
I don't think Node.js considers docs in source code as promised behavior.

Reply:
I'm also unclear on how `It means 'readable'/'data' event is emitted immediately or not.` has any indication on if the http/2 stream has data frames or not? It seems to be unrelated to the unlying state of the http/2 layer to me.

Reply:
Searching the Node.js issue tracker for the underscore property pulls up https://github.com/nodejs/node/issues/445 which sounds like it shouldn't be used inside Node.js code, let alone third-party things.

Reply:
> I don't think Node.js considers docs in source code as promised behavior.

Yes. However, current Nodejs has no other parameter to indicate hasBody after event 'end'. Should we suggest new API for Nodejs?

> I'm also unclear on how It means 'readable'/'data' event is emitted immediately or not. has any indication on if the http/2 stream has data frames or not? It seems to be unrelated to the unlying state of the http/2 layer to me.

`sync` relates to readableStream rather than http/2 layer state. When http/2 layer get DataFrame, readableStream(ServerHttp2Request) emit 'data' event immediately or not.

Reply:
> Yes. However, current Nodejs has no other parameter to indicate hasBody after event 'end'. Should we suggest new API for Nodejs?

Based on the discussion in https://github.com/nodejs/node/issues/445 it sounds like the answer is it's a must, as the activity seems to indicate that this will break sooner rather than later, and they want to get everything public.

Reply:
OK, I will file a issue in nodejs repo and pending this PR.